rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+---
+feature: pipe-operator
+start-date: 2023-05-23
+author: @piegamesde
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a new "pipe" operator, `|>`, to the Nix language, defined as `f a` = `a |> f`.
+Additionally, elevate `lib.pipe` to a built-in function.
+
+As a reminder, `pipe a [ f g h ]` is defined as `h (g (f a))`.
+
+# Motivation
+[motivation]: #motivation
+
+Creating advanced data processing like transforming a list is a thing commonly done in nixpkgs.
+Yet the language has no support for function concatentation/composition,
+which results in such constructs looking unwieldy and difficult to format well.
+`lib.pipe` may be the most powerful library function with that regard,
+but it is unknown and overlooked by many because it is not easily discoverable:
+Despite its great usefulness, it is currently used in less than 30 files in Nixpkgs
```

Comment:
I don't think I've even see `lib.pipe` anywhere. I'm guessing that it is the case for A LOT of people.

I think this argument says much more about the state of the documentation and the bizarre `builtins` vs `nixpkgs.lib` vs `nixos-lib` fenomena we have.

I think this would become a non-issue if `nixpkgs.lib` contained only functions that did something internal to Nixpkgs specifically (such as `lib.options`), and everything that manipulates data like strings, lists and attrsets was moved to `builtins` and folded nicely. By "folding" I mean that I think `map`, `mapAttrs`, `mapAttrs` flavors and `mapNullable` should exist in the same section in the manual. If I know how `map` works in the general sense, then I'll be able to understand how `mapAttrs` works. I would then be able to see all of the types `map` is applicable on and the different variations of the function by clicking on the "`map` function" header.

That said, this is outside of the scope of this RFC, so I digress. I do think that the issue lies in the documentation and technical structure though, and it would be nice if we started talking about that.

Reply:
Shepherds currently believe this conversation is resolved, outcome: mostly irrelevant to this RFC.